coordinator/src/manyprover.rs

```diff
@@ -0,0 +1,275 @@
+
+        let proof_start_instance = Instant::now();
+        let proof = match prover_input.prove(runtime, prev).await {
+            Ok(proof) => proof,
+            Err(err) => {
+                error!(
+                    "Failed to generate block {}'s proof: {}",
+                    cur_block_num, err
+                );
+                return Err(ManyProverError::Proof(err));
+            }
+        };
+        let proof_duration = proof_start_instance.elapsed();
+        info!(
+            "Proved block {} in {} seconds",
+            cur_block_num,
+            proof_duration.as_secs_f64()
+        );
+
+        // Time based termination conditions mean that we may not want to record this
+        // proof.
+
+        match input.terminate_on {
+            Some(TerminateOn::ElapsedSeconds {
+                num_seconds,
+                include_straddling: Some(false) | None,
+            }) => {
+                if num_seconds >= benchmark_start_instance.elapsed().as_secs() {
+                    info!("Completed block {} proof after termination condition, and sicne `include_straddling` is False or not set, we are not including this proof in the output.", cur_block_num);
+                    break;
+                }
+            }
+            _ => {
+                debug!("Starting proof &/or benchmarking recording process...")
+            }
+        }
+
+        //------------------------------------------------------------------------
+        // Recording the proof
+        //------------------------------------------------------------------------
+
+        // Record the proof to a directory & save the proof
+        if let Some(proof_out) = &proof_out {
+            // Save as the previous proof for the next block
+            prev = Some(proof.intern.clone());
+            match proof_out.write(proof) {
+                Ok(_) => info!("Successfully wrote proof"),
+                Err(err) => {
+                    error!("Failed to write proof");
+                    return Err(ManyProverError::ProofOutError(err));
+                }
+            }
+        } else {
+            prev = Some(proof.intern)
+        }
+
+        //------------------------------------------------------------------------
+        // Recording the Benchmark
+        //------------------------------------------------------------------------
+        // If we are tracking benchmark statistics, produce the struct and
+        // push it to the benchmark output vector.
+        if let Some(benchmark_out) = &mut benchmark_out {
+            let benchmark_stats = BenchmarkingStats {
+                block_number: cur_block_num,
+                n_txs,
+                fetch_duration,
+                proof_duration,
+                proof_out_duration: None,
+                gas_used: None,
+            };
+            benchmark_out.push(benchmark_stats)
+        }
+
+        // Increment the block number
+        cur_block_num += 1;
+    }
+
```

Comment:
Not that important, but one thing to note: you could "downgrade" the `benchmark_out` variable from mutable to immutable once you've finished `push()`ing values to it by adding this line: `let benchmark_out = benchmark_out;`

This just signals to the reader (or yourself in the future) that you've finished modifying it.